Fix part of #13822: Write some beam jobs for exploration validation

soumyo123-prog opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #13822 
2. This PR writes beam jobs for:
- Exploration param_specs and param_changes should be empty for curated lessons.
- State classifier model_id should be None for curated lessons.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes 
made in this PR work correctly. If this PR is for a developer-facing feature, 
provide the videos/screenshots of developer-facing interface. 

Please also include videos/screenshots of the developer tools 
browser console, so that we can be sure that there are no console errors.

If the changes in your PRs are autogenerated via a script and you cannot provide proof 
for the changes then please leave a comment "No proof of changes needed because {{Reason}}" 
and remove all the sections below.
-->





![Screenshot from 2022-05-05 15-50-17](https://user-images.githubusercontent.com/70739349/166904865-e3be7616-4d5d-46be-81a0-a8e31dd923f8.png)


![Screenshot from 2022-05-05 15-50-19](https://user-images.githubusercontent.com/70739349/166904873-df53e19d-ee18-476d-ab90-0e07c2b49b35.png)

#### Proof of changes on desktop with slow/throttled network

<!--
Make sure to properly verify that everything works correctly, and that there are 
no weird UI mistakes or other problems. Also, if there are any newly added fields, 
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below). 
There should be no performance or UI issues while the network is slow.

References: 
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to 
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI in any of the files listed in rtl_css.py (i.e, those that have 
a separate .rtl.css file for styling), make sure to add screenshots with the site 
language set to Arabic as well (we use Arabic as it is a language written from right to left).  
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

soumyo123-prog committed:
written the job

soumyo123-prog committed:
wrote the second job

soumyo123-prog committed:
wrote the job for exp outcome validation

soumyo123-prog committed:
fix lint errors

vojtechjelinek commented:
@soumyo123-prog Do you want a review?

soumyo123-prog commented:
> @soumyo123-prog Do you want a review?

@vojtechjelinek Not currently, I will request for review once it is completed.

soumyo123-prog committed:
merge remote branch oppia/develop

soumyo123-prog committed:
fixing

soumyo123-prog committed:
fix lint errors

soumyo123-prog committed:
added for training data

soumyo123-prog committed:
fix lint errors

soumyo123-prog committed:
fix lint errors

soumyo123-prog committed:
fixing

soumyo123-prog commented:
@vojtechjelinek PTAL. Thanks!

vojtechjelinek commented:
@lkbhitesh07 Please take the first pass.

lkbhitesh07 reviewed (commented):
Hey @soumyo123-prog, Thanks for the PR.
Actually, the reason why we were combining some similar checks is to save some complexity so that the load can be reduced and here you are initializing pipeline several times getting the "ExplorationModel", I think you need to combine all the checks.

Also here you are only checking for `explorations` but we need to check for curated lessons which means the explorations which are present or attached to `chapters`.
Thanks

soumyo123-prog committed:
temp

soumyo123-prog committed:
small fix

soumyo123-prog committed:
fix lint errors - 1

soumyo123-prog committed:
fix lint errors - 2

soumyo123-prog committed:
merge remote develop into compiled_bj

soumyo123-prog commented:
@lkbhitesh07 PTAL! I have combined the checks now. Thanks!

lkbhitesh07 commented on core/jobs/batch_jobs/exp_validation_jobs_test.py:
```diff
@@ -0,0 +1,388 @@
+            }
+        )
+
+        self.opportunity_1 = self.create_model(
```
It will be helpful to have a comment above regarding valid or invalid, same as you have done in explorations.
Thanks

soumyo123-prog replied:
Here, all the models are valid. These are just added so that the explorations become curated.

soumyo123-prog replied:
I have added the comment saying that all the opportunity models are valid.

lkbhitesh07 commented on core/jobs/batch_jobs/exp_validation_jobs.py:
```diff
@@ -0,0 +1,261 @@
+          defined state classifier model id.
+      b) The exploration is curated and has non-empty param changes.
+      c) The exploration is curated and has non-empty param specs.
+      d) The exploration is curated and has non-empty training data.
```
If you want you can combine last three into one, totally your call.
Thanks

soumyo123-prog replied:
I just noticed that the 4th docstring was misleading. I have updated it now.

lkbhitesh07 reviewed (changes requested):
Great work, just some basic nits. Thanks for the changes @soumyo123-prog.

soumyo123-prog committed:
Merge remote-tracking branch 'upstream/develop' into compiled_bj

soumyo123-prog committed:
small nits

soumyo123-prog committed:
fix lint errors

soumyo123-prog committed:
fix lint errors 2

soumyo123-prog commented:
@lkbhitesh07 PTAL. Thanks!

soumyo123-prog commented:
@vojtechjelinek I think @lkbhitesh07 's review is almost over now. Could you please take a look on the changes? Thanks!

lkbhitesh07 reviewed (approved):
Looks good to me
Thanks

vojtechjelinek commented on core/jobs/batch_jobs/exp_validation_jobs.py:
```diff
@@ -0,0 +1,280 @@
+
+(exp_models, opportunity_models) = models.Registry.import_models([
+    models.NAMES.exploration,
+    models.NAMES.opportunity
```
```suggestion
    models.NAMES.exploration, models.NAMES.opportunity
```

soumyo123-prog replied:
Done

vojtechjelinek commented on core/jobs/batch_jobs/exp_validation_jobs.py:
```diff
@@ -0,0 +1,280 @@
+        for state_name in states:
+            if states[state_name].classifier_model_id is not None:
+                return True
+        return False
```
```suggestion
        return any(
            state.classifier_model_id is not None for state in states.values()
        )
```

soumyo123-prog replied:
Done

vojtechjelinek commented on core/jobs/batch_jobs/exp_validation_jobs.py:
```diff
@@ -0,0 +1,280 @@
+            opportunity_model = (
+                opportunity_models.ExplorationOpportunitySummaryModel.get(
+                    exploration.id, strict=False))
+        return opportunity_model is not None
```
Is this a valid way of checking curated exploration? Also, I think we should rather get all the `ExplorationOpportunitySummaryModel` and then use filter with side values. (But there might be other models that might be used for this).

soumyo123-prog replied:
I think this is a valid way for checking curated exploration. In the `opportunity_services.is_exploration_available_for_contribution`, curated exploration is checked in this way.

vojtechjelinek replied:
The other part of my comment is not addressed.

soumyo123-prog replied:
Ah I missed that somehow. I am working on it now. Thanks!

soumyo123-prog replied:
It is addressed now. PTAL. Thanks!

vojtechjelinek replied:
Sorry, the solution is not correct. You will need to fetch the models using `GetModels`. Similar to what you do with `ExplorationModel`s.

vojtechjelinek commented on core/jobs/batch_jobs/exp_validation_jobs.py:
```diff
@@ -0,0 +1,280 @@
+            ].interaction.answer_groups:
+                if len(answer_group.training_data) > 0:
+                    state_names.append(state_name)
+        return state_names
```
Could these two be done in one step?

soumyo123-prog replied:
Done

vojtechjelinek commented on core/jobs/batch_jobs/exp_validation_jobs.py:
```diff
@@ -0,0 +1,280 @@
+
+        report_number_of_exps_having_invalid_state_classifer_model_id = (
+            curated_exps_having_invalid_state_classifier_model_id
+            | 'Count explorations having invalid state classifier model id' >>
+                (job_result_transforms.CountObjectsToJobRunResult(
+                    'INVALID STATE CLASSIFIER'))
```
```suggestion
            | 'Count explorations having invalid state classifier model id' >>
                job_result_transforms.CountObjectsToJobRunResult(
                    'INVALID STATE CLASSIFIER')
```

soumyo123-prog replied:
Done

vojtechjelinek commented on core/jobs/batch_jobs/exp_validation_jobs.py:
```diff
@@ -0,0 +1,280 @@
+
+        curated_exps_having_invalid_state_classifier_model_id = (
+            curated_explorations
+            | 'Filter curated explorations having invalid state classifier '
+            'model id' >>
+                beam.Filter(lambda exp: (
+                    self.filter_exps_having_invalid_state_classifier_model_id(
+                        exp.states
+                    )))
+        )
```
```suggestion
            | 'Filter curated exps having invalid state classifier model id' >>
                beam.Filter(lambda exp: (
                    self.filter_exps_having_invalid_state_classifier_model_id(
                        exp.states
                    )
                ))
        )
```

soumyo123-prog replied:
done

vojtechjelinek commented on core/jobs/batch_jobs/exp_validation_jobs.py:
```diff
@@ -0,0 +1,280 @@
+            PCollection[JobRunResult]. A PCollection of details of explorations
+            which are invalid.
+        """
+        total_explorations = (
+            self.pipeline
+            | 'Get all ExplorationModels' >> ndb_io.GetModels(
+                exp_models.ExplorationModel.get_all(include_deleted=False))
+            | 'Get exploration from model' >> beam.Map(
+                exp_fetchers.get_exploration_from_model)
+        )
```
```suggestion
        all_explorations = (
            self.pipeline
            | 'Get all ExplorationModels' >> ndb_io.GetModels(
                exp_models.ExplorationModel.get_all(include_deleted=False))
            | 'Get exploration from model' >> beam.Map(
                exp_fetchers.get_exploration_from_model)
        )
```

soumyo123-prog replied:
done

vojtechjelinek reviewed (changes requested):
Thanks! Took a pass.

soumyo123-prog committed:
Merge remote-tracking branch 'upstream/develop' into compiled_bj

soumyo123-prog committed:
some nits

soumyo123-prog committed:
fixes

soumyo123-prog committed:
fix lint errors

soumyo123-prog commented:
@vojtechjelinek I have answered the comments. PTAL. Thanks!

soumyo123-prog committed:
improvements

soumyo123-prog committed:
fix lint errors

soumyo123-prog commented:
@vojtechjelinek I have made the changes and fetched all the opportunity models beforehand. PTAL. Thanks!

soumyo123-prog committed:
fixes

soumyo123-prog commented:
@vojtechjelinek I have made the changes. PTAL. Thanks!

soumyo123-prog committed:
Merge remote-tracking branch 'upstream/develop' into compiled_bj

vojtechjelinek commented on core/jobs/batch_jobs/exp_validation_jobs.py:
```diff
@@ -0,0 +1,298 @@
+        Returns:
+            Exploration. The exploration domain object.
+        """
+        exp_and_opportunity_models = list(model_pair[1])
+        exp_model = exp_and_opportunity_models[0]
+        if not isinstance(exp_model, exp_models.ExplorationModel):
+            exp_model = exp_and_opportunity_models[1]
+        return exp_fetchers.get_exploration_from_model(exp_model)
```
This is a bit iffy. Can we somehow make sure that the tuple will always have the same order?

soumyo123-prog replied:
I have done this now by COGroupByKey. PTAL.

vojtechjelinek commented on core/jobs/batch_jobs/exp_validation_jobs.py:
```diff
@@ -0,0 +1,298 @@
+                    break
+        return state_names
+
+    def filter_curated_explorations(self, model_pair):
+        """Returns True if the exploration is curated.
+
+        Args:
+            model_pair: tuple. The pair of exp and opportunity models.
+
+        Returns:
+            bool. Returns True if opportinity model exists for the given id.
+        """
+        exp_and_opportunity_models = list(model_pair[1])
+        return len(exp_and_opportunity_models) == 2
```
```suggestion
    def filter_curated_explorations(self, model_pair):
        """Returns True if the exploration is curated.

        Args:
            model_pair: tuple. The pair of exp and opportunity models.

        Returns:
            bool. Returns True if opportinity model exists for the given id.
        """
        return len(list(model_pair[1])) == 2
```

soumyo123-prog replied:
Done

vojtechjelinek commented on core/jobs/batch_jobs/exp_validation_jobs.py:
```diff
@@ -0,0 +1,298 @@
+    models.NAMES.exploration, models.NAMES.opportunity])
+
+
+class GetNumberOfInvalidExplorationsJob(base_jobs.JobBase):
```
Please add typing for everything in this file. It will make reviewing easier.

soumyo123-prog replied:
Added the typings.

vojtechjelinek commented on core/jobs/batch_jobs/exp_validation_jobs.py:
```diff
@@ -0,0 +1,298 @@
+        """
+        all_explorations = (
+            self.pipeline
+            | 'Get all ExplorationModels' >> ndb_io.GetModels(
+                  exp_models.ExplorationModel.get_all(include_deleted=False))
```
```suggestion
            | 'Get all ExplorationModels' >> ndb_io.GetModels(
                exp_models.ExplorationModel.get_all(include_deleted=False))
```

soumyo123-prog replied:
Done

vojtechjelinek commented on core/jobs/batch_jobs/exp_validation_jobs_test.py:
```diff
@@ -0,0 +1,394 @@
+            state_domain.RuleSpec(
+                'Equals', {
+                    'x': {
+                        'contentId': 'rule_input_Equals',
+                        'normalizedStrSet': ['Test']
+                        }})
```
```suggestion
                        'contentId': 'rule_input_Equals',
                        'normalizedStrSet': ['Test']
                    }
                }
            )
```

vojtechjelinek replied:
ditto elsewhere

soumyo123-prog replied:
Done

vojtechjelinek commented on core/jobs/batch_jobs/exp_validation_jobs_test.py:
```diff
@@ -0,0 +1,394 @@
+    def test_run_with_no_models(self) -> None:
+        self.assert_job_output_is([])
+
+    def test_run_with_single_valid_model(self) -> None:
+        self.put_multi([self.exp_1, self.opportunity_1])
+        self.assert_job_output_is([
+            job_run_result.JobRunResult.as_stdout('EXPS SUCCESS: 1'),
+            job_run_result.JobRunResult.as_stdout('CURATED EXPS SUCCESS: 1')
+        ])
```
Can you add test with non-curated exploration?

soumyo123-prog replied:
Added that

vojtechjelinek commented on core/jobs/batch_jobs/exp_validation_jobs_test.py:
```diff
@@ -0,0 +1,394 @@
+            job_run_result.JobRunResult.as_stdout('CURATED EXPS SUCCESS: 1')
+        ])
+
+    def test_run_with_single_invalid_model(self) -> None:
```
Maybe have a separate test for each invalidity type?

soumyo123-prog replied:
Done

vojtechjelinek reviewed (changes requested):
Thanks! Took a pass.

soumyo123-prog committed:
add typing in the file

soumyo123-prog committed:
added test for non-curated exp

soumyo123-prog committed:
added tests for individual invalidity types

soumyo123-prog committed:
fix lint errors

soumyo123-prog committed:
improvement

soumyo123-prog committed:
fix lint errors

soumyo123-prog commented:
@vojtechjelinek PTAL. Addressed your comments. Thanks!

soumyo123-prog committed:
Merge remote-tracking branch 'upstream/develop' into compiled_bj

soumyo123-prog commented:
@vojtechjelinek Can I submit the testing request?

vojtechjelinek commented on core/jobs/batch_jobs/exp_validation_jobs.py:
```diff
@@ -0,0 +1,311 @@
+          interaction having answer group(s) with non-empty training data.
+    """
+
+    def get_states_having_invalid_state_classifier_id(
+        self, states: dict[str, state_domain.State]) -> list[str]:
```
```suggestion
    def get_states_having_invalid_state_classifier_id(
        self, states: Dict[str, state_domain.State]
    ) -> List[str]:
```

ditto below

vojtechjelinek commented on core/jobs/batch_jobs/exp_validation_jobs.py:
```diff
@@ -0,0 +1,311 @@
+            list[str]. Returns a list of state names that have not None
+            classifier model id.
+        """
+        invalid_state_names: list[str] = []
```
```suggestion
        invalid_state_names: List[str] = []
```

soumyo123-prog replied:
Done.

vojtechjelinek commented on core/jobs/batch_jobs/exp_validation_jobs.py:
```diff
@@ -0,0 +1,311 @@
+        return state_names
+
+    def filter_curated_explorations(
+        self, exp_and_opp_models: tuple[str, dict[str, list[object]]]):
```
Why `object`?

soumyo123-prog replied:
Modified it now.

vojtechjelinek commented on core/jobs/batch_jobs/exp_validation_jobs.py:
```diff
@@ -0,0 +1,311 @@
+        return len(exp_models_list) == 1 and len(opp_models_list) == 1
+
+    def get_exploration_from_models(
+        self, exp_and_opp_models: tuple[str, dict[str, list[object]]]):
```
Again, why object, also missing return type.

soumyo123-prog replied:
Modified it now.

vojtechjelinek commented on core/jobs/batch_jobs/exp_validation_jobs.py:
```diff
@@ -0,0 +1,311 @@
+                'exp_opportinity_model': all_exp_opportunities
+            })
+            | 'Combine the PCollections' >> beam.CoGroupByKey()
+            | 'Filter curated explorations' >> beam.Filter(
+                self.filter_curated_explorations)
```
Maybe you can also reformat the pcoll returned by `CoGroupByKey` so that it is just a dictionary with key and model value. Because now you need to do `list(exp_and_opp_models[1]['exp_model'])` above, yo should modify it here so that you can do just `exp_and_opp_models['exp_model']` above to get the model.

soumyo123-prog replied:
I just noticed that we don't need to convert into list when we use `CoGroupByKey`. I have modified this functionality now. PTAL. Thanks!

vojtechjelinek commented on core/jobs/batch_jobs/exp_validation_jobs.py:
```diff
@@ -0,0 +1,311 @@
+            PCollection[JobRunResult]. A PCollection of details of explorations
+            which are invalid.
+        """
+        all_explorations: beam.PCollection = (
```
You don't need to add the types here. Only add types to places where MyPy checks fail for this file.

vojtechjelinek replied:
ditto below

soumyo123-prog replied:
Removed them

vojtechjelinek reviewed (changes requested):
Thanks! Took a pass.

soumyo123-prog committed:
changes in typing and improvements

soumyo123-prog committed:
fixed lint errors

soumyo123-prog commented:
@vojtechjelinek I have made the changes as per your comments. PTAL. Thanks!

vojtechjelinek commented on core/jobs/batch_jobs/exp_validation_jobs.py:
```diff
@@ -0,0 +1,350 @@
+                'exp_models': all_explorations,
+                'opportinity_models': all_exp_opportunities
+            })
+            | 'Combine the PCollections' >> beam.CoGroupByKey()
```
You can also pass tuple to `CoGroupByKey`, that **might** simplify some of the code below.
```suggestion
            (all_explorations, all_exp_opportunities)
            | 'Combine the PCollections' >> beam.CoGroupByKey()
```

soumyo123-prog replied:
Done.

vojtechjelinek commented on core/jobs/batch_jobs/exp_validation_jobs.py:
```diff
@@ -0,0 +1,350 @@
+
+
+class ExpAndOpportunityModelsDict(TypedDict):
+    """Dictionary that represents each element of the PCollection returned by
+    CoGroupByKey."""
```
```suggestion
    """Dictionary representing the elements returned by CoGroupByKey."""
```

soumyo123-prog replied:
Removed this portion now as it is no longer needed.

vojtechjelinek commented on core/jobs/batch_jobs/exp_validation_jobs.py:
```diff
@@ -0,0 +1,350 @@
+    CoGroupByKey."""
+
+    exp_models: List[exp_models.ExplorationModel]
+    opportinity_models: List[
```
```suggestion
    opportunity_models: List[
```

soumyo123-prog replied:
Removed this portion now as it is no longer needed.

vojtechjelinek reviewed (changes requested):
Thanks! Left a few comments; after that, you can submit the backup server testing request.

soumyo123-prog committed:
some nits

soumyo123-prog committed:
Merge remote-tracking branch 'upstream/develop' into compiled_bj

soumyo123-prog commented:
@vojtechjelinek I have submitted the request.

vojtechjelinek commented:
@soumyo123-prog The results of this job are:
```
EXPS SUCCESS: 35358
CURATED EXPS SUCCESS: 61
```

soumyo123-prog commented:
Closing this PR now as all explorations are passing. I will raise validation PR after discussing some things with others.

soumyo123-prog committed:
Merge remote-tracking branch 'upstream/develop' into compiled_bj

soumyo123-prog committed:
added some more checks

soumyo123-prog committed:
lint errors fix

soumyo123-prog committed:
temp

soumyo123-prog committed:
temp

vojtechjelinek commented:
@soumyo123-prog Is this ready for review?

soumyo123-prog commented:
> @soumyo123-prog Is this ready for review?

@vojtechjelinek I had to reopen this because I had to add 3 new checks on this PR itself. I will make ready for review asap. I will reassign you once I finish this.

soumyo123-prog committed:
all fixed

soumyo123-prog committed:
Merge remote-tracking branch 'upstream/develop' into compiled_bj

soumyo123-prog committed:
fix lint errors

soumyo123-prog commented:
@vojtechjelinek PTAL. This job should be run again for checking the new validations added.

soumyo123-prog committed:
add docstring

vojtechjelinek commented on core/jobs/batch_jobs/exp_validation_jobs.py:
```diff
@@ -0,0 +1,503 @@
+          interaction having answer group(s) with non-empty training data.
+      e) The exploration is curated and has any state with MultipleChoiceInput
+          interaction having less than 4 choices.
+      f) The exploration is curated and any state with default outcome where
+          param changes is not empty.
```
I don't understand this sentence.

soumyo123-prog replied:
I have modified this now. PTAL. Thanks!

vojtechjelinek commented on core/jobs/batch_jobs/exp_validation_jobs_test.py:
```diff
@@ -0,0 +1,884 @@
+            )
+        ])
+
+    def test_run_with_invalid_cust_args(self) -> None:
```
Can you also test all the other problems separately?

soumyo123-prog replied:
The individual tests for all the other problems are already added.

vojtechjelinek reviewed (changes requested):
Thanks! Just two small notes.

soumyo123-prog committed:
small nit

soumyo123-prog commented:
@vojtechjelinek PTAL. Thanks!

vojtechjelinek commented:
@soumyo123-prog Looks good. I will test this before Wednesday EOD.

vojtechjelinek commented:
@soumyo123-prog This will probably be done tomorrow.

vojtechjelinek commented:
@soumyo123-prog Sorry, this will take some time as I found a bug in our package code that is used by jobs.

soumyo123-prog committed:
fix the job output

soumyo123-prog committed:
Merge remote-tracking branch 'upstream/develop' into compiled_bj

soumyo123-prog commented:
@vojtechjelinek This job needs to be run once again as I have changed the output for videos or links a little bit.

vojtechjelinek committed:
Merge branch 'develop' into compiled_bj

vojtechjelinek commented:
@soumyo123-prog I will try to run this on Monday.

vojtechjelinek commented:
@soumyo123-prog This fails on the backup server:
```
File "core/jobs/batch_jobs/exp_validation_jobs.py", line 248, in get_exploration_from_models 
  return exp_fetchers.get_exploration_from_model(model_pair[0]) 
File "core/domain/exp_fetchers.py", line 197, in get_exploration_from_model 
  _migrate_states_schema(versioned_exploration_states, init_state_name) 
File "core/domain/exp_fetchers.py", line 93, in _migrate_states_schema 
  states_schema_version, init_state_name) 
File "core/domain/exp_domain.py", line 2238, in update_states_from_model 
  versioned_exploration_states['states'])
File "core/domain/exp_domain.py", line 2132, in _convert_states_v47_dict_to_v48_dict 
  state_schema_version=48) 
File "core/domain/state_domain.py", line 3968, in convert_html_fields_in_state 
  conversion_fn 
File "core/domain/state_domain.py", line 1067, in convert_html_in_interaction 
  ca_specs_dict) 
File "core/domain/state_domain.py", line 1464, in convert_cust_args_dict_to_cust_args_based_on_specs 
  ) for spec in ca_specs_dict
File "core/domain/state_domain.py", line 1464, in <dictcomp> 
  ) for spec in ca_specs_dict 
KeyError: "allowedVariables [while running 'Get exploration from model-ptransform-301']" 
```

soumyo123-prog committed:
Merge remote-tracking branch 'upstream/develop' into compiled_bj

soumyo123-prog committed:
minor hacky solution

soumyo123-prog committed:
fix mypy

soumyo123-prog committed:
Merge branch 'compiled_bj' of https://github.com/soumyo123-prog/oppia into compiled_bj

soumyo123-prog commented:
> @soumyo123-prog This fails on the backup server:
> 
> ```
> File "core/jobs/batch_jobs/exp_validation_jobs.py", line 248, in get_exploration_from_models 
>   return exp_fetchers.get_exploration_from_model(model_pair[0]) 
> File "core/domain/exp_fetchers.py", line 197, in get_exploration_from_model 
>   _migrate_states_schema(versioned_exploration_states, init_state_name) 
> File "core/domain/exp_fetchers.py", line 93, in _migrate_states_schema 
>   states_schema_version, init_state_name) 
> File "core/domain/exp_domain.py", line 2238, in update_states_from_model 
>   versioned_exploration_states['states'])
> File "core/domain/exp_domain.py", line 2132, in _convert_states_v47_dict_to_v48_dict 
>   state_schema_version=48) 
> File "core/domain/state_domain.py", line 3968, in convert_html_fields_in_state 
>   conversion_fn 
> File "core/domain/state_domain.py", line 1067, in convert_html_in_interaction 
>   ca_specs_dict) 
> File "core/domain/state_domain.py", line 1464, in convert_cust_args_dict_to_cust_args_based_on_specs 
>   ) for spec in ca_specs_dict
> File "core/domain/state_domain.py", line 1464, in <dictcomp> 
>   ) for spec in ca_specs_dict 
> KeyError: "allowedVariables [while running 'Get exploration from model-ptransform-301']" 
> ```

@vojtechjelinek This looks strange as in the last commit, I just changed the structure of the output and did not touch any of the code responsible for the transform 'Get exploration from model' and the job did run successfully multiple times in the server.

Anyways, I have now added a safeguard that if get_exploration_from_model does not run properly, then ignore that exploration. PTAL. Thanks!

vojtechjelinek commented on core/jobs/batch_jobs/exp_validation_jobs.py:
```diff
@@ -245,7 +245,13 @@ def get_exploration_from_models(self, model_pair: Tuple[
+            )
+            return exploration
+        except Exception:
+            return None
```
I will run the job but this is not ideal, you should record how many explorations are skipped.

soumyo123-prog replied:
I have incorporated it now but as the results are already here, do we still need that?

Also, just curious that "can beam job runs flake sometimes"?

vojtechjelinek replied:
@soumyo123-prog I think there isn't anything that cannot flake, but I've not yet Beam job flake. It usually successfully completed or failed for some reason.

vojtechjelinek commented:
@soumyo123-prog Results are on your email.

soumyo123-prog committed:
some improvement on the output

soumyo123-prog commented:
> @soumyo123-prog Results are on your email.

Thanks and I will take care of that from next time.

soumyo123-prog commented:
Closing this PR for now as everything seems to be done about this PR.

lkbhitesh07 commented:
Hey @soumyo123-prog, Just wanted to get a follow-up on this PR. Why was this PR closed?
